lib/metadata/chordMetadata.hpp

```diff
@@ -164,50 +176,43 @@ struct chordMetadata {
     }
 };
 
-inline void chord_metadata_init(chordMetadata* c) {
-    bzero(c, sizeof(chordMetadata));
+inline void chord_metadata_init(std::shared_ptr<chordMetadata> meta) {
+    // ??
+    *meta = chordMetadata();
 }
 
-inline void chord_metadata_copy(chordMetadata* out, const chordMetadata* in) {
-    memcpy(out, in, sizeof(chordMetadata));
+inline void chord_metadata_copy(std::shared_ptr<chordMetadata> out,
+                                const std::shared_ptr<chordMetadata> in) {
+    *out = *in; // ???
 }
 
 inline bool metadata_is_chord(Buffer* buf, int) {
-    return strcmp(buf->metadata_pool->type_name, "chordMetadata") == 0;
+    return buf && buf->metadata_pool && (buf->metadata_pool->type_name == "chordMetadata");
 }
 
-inline bool metadata_container_is_chord(const metadataContainer* mc) {
-    return strcmp(mc->parent_pool->type_name, "chordMetadata") == 0;
-}
-
-inline const chordMetadata* get_chord_metadata(const Buffer* buf, int frame_id) {
-    return (const chordMetadata*)buf->metadata[frame_id]->metadata;
-}
-
-inline chordMetadata* get_chord_metadata(Buffer* buf, int frame_id) {
-    return (chordMetadata*)buf->metadata[frame_id]->metadata;
+inline bool metadata_is_chord(const std::shared_ptr<metadataObject> mc) {
+    if (!mc)
+        return false;
+    std::shared_ptr<metadataPool> pool = mc->parent_pool.lock();
+    assert(pool);
+    return (pool->type_name == "chordMetadata");
 }
 
-inline const chordMetadata* get_chord_metadata(const metadataContainer* mc) {
-    if (!mc)
-        return nullptr;
-    if (strcmp(mc->parent_pool->type_name, "chordMetadata")) {
-        WARN_NON_OO("Expected metadata to be type \"chordMetadata\", got \"{:s}\".",
-                    mc->parent_pool->type_name);
-        return nullptr;
-    }
-    return static_cast<const chordMetadata*>(mc->metadata);
+inline std::shared_ptr<chordMetadata> get_chord_metadata(Buffer* buf, int frame_id) {
+    std::shared_ptr<metadataObject> meta = buf->metadata[frame_id];
```

Comment:
Should there be an assert statement checking that the buffer has metadata and that the metadata is in chord format?

Reply:
Good catch, done!